### ISSUE TITLE
Setting labels in unmanaged views

### ISSUE THREAD

**Issue by chmaruni** [kokkos/kokkos#1963]
As I understand it, view labels are essentially used for debugging/profiling. However, unmanaged views apparently cannot have labels, according to a comment in the unmanaged views code example on the Kokkos wiki. Is there a good reason for this behavior? Or is there some way to set the label of an unmanaged view? 

We would need labels for unmanaged views because we would like to store views into files with their associated string name.

**Comment by chmaruni** [comment]
Ah, so is the idea that an unmanaged view uses zero memory and only exists as a collection of templated methods at compile time? I assumed a view would be essentially something like { header_type header; ptr_type memory; some_other stuff; } but you are saying that without a header there is only the pointer and nothing else and the view doesn't really exist at run time?

That makes sense, but maybe you have a suggestion how I could best add a label to an unmanaged view? If I take a subview of the unmanaged view, will that have a header with a label?

**Comment by mhoemmen** [comment]
Why can’t you just make up a label for unmanaged Views?  You could generate unique names with a counter.

**Comment by chmaruni** [comment]
I don't want a random name but a specific one. Essentially, in one part of the code I want to create unmanaged views like `auto p = Kokkos::View("pressure_field", some_array, some_strided_layout);`

Then later I would like to pass this view do a function like `dump_to_netcdf(p);` where I would like to use the label to name the array in the netcdf file.

**Comment by chmaruni** [comment]
True, and that's what I will do if there is no other way. I just thought that maybe I can avoid it since a Kokkos::View already has a label, and that's really all I need. 

I guess, given that labels are (according to the doc) meant for improving debugging and profiling output I just don't understand why they are a feature of the memory tracking? I want meaningful debugging and profiling for unmanaged views as well, not only for normal views, no?

**Comment by dsunder** [comment]
Adding a label to unmanaged views would add **significant** overhead.

Managed views in host memory require one allocation that reserves space for a header before the view to store the reference count for the view and the label.  A managed view in GPU memory requires both a host allocation for the reference count and a GPU allocation where we reserve space for a header that stores the host pointer for the view and the label.

Unmanaged views are designed to have near zero overhead that work with any pointer that the user provides and can be created inside of parallel algorithms.  There is no space to store a label.  If we wanted to be able to add a label to an unmanaged view we would need to use something like an lockfree unordered_map to map from the pointer to the label, which would significantly degrade the performance and usefulness of unmanaged views.

TLDR; If you want a label you **must** use managed views.

**Comment by chmaruni** [comment]
Makes sense, thanks, that's what I figured. Thanks for the confirmation and additional info. Closing the issue.
